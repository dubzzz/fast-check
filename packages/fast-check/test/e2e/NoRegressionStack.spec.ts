import fc from '../../src/fast-check';

const settings = { seed: 42, verbose: 0 };

describe(`NoRegressionStack`, () => {
  it('throw', () => {
    expect(
      sanitize(() =>
        fc.assert(
          fc.property(fc.nat(), fc.nat(), (a, b) => {
            if (a < b) {
              throw new Error('a must be >= b');
            }
          }),
          settings,
        ),
      ),
    ).toThrowErrorMatchingSnapshot();
  });

  it('not a function', () => {
    expect(
      sanitize(() =>
        fc.assert(
          fc.property(fc.nat(), (v) => {
            (v as any)();
          }),
          settings,
        ),
      ),
    ).toThrowErrorMatchingSnapshot();
  });
});

// Helpers

function sanitize(run: () => void) {
  return () => {
    try {
      run();
    } catch (err) {
      const initialMessage = (err as Error).message;
      const lines = initialMessage
        .replace(/\\/g, '/')
        .replace(/at [^(]*fast-check\/(packages|node_modules)(.*)/g, 'at $1$2')
        .replace(/at (.*) \(.*fast-check\/(packages|node_modules)(.*)\)/g, 'at $1 ($2$3)')
        .replace(/at (.*) \(.*\/(\.yarn|Yarn)\/.*\/(node_modules\/.*):\d+:\d+\)/g, 'at $1 ($3:?:?)') // reducing risks of changes on bumps: .yarn (Linux and Mac), Yarn (Windows)
        .split('\n');
      throw new Error(
        lines
          .slice(
            0,
            // internals of jest, subject to regular changes
            // and OS dependent
            lines.findIndex((line) => line.includes('node_modules/jest-circus')),
          )
          .join('\n'),
      );
    }
  };
}
