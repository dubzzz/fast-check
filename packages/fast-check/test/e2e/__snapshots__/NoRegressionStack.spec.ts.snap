// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoRegressionStack not a function 1`] = `
"Property failed after 1 tests
{ seed: 42, path: "0:0", endOnFailure: true }
Counterexample: [0]
Shrunk 1 time(s)
Got TypeError: v is not a function
    at v (packages/fast-check/test/e2e/NoRegressionStack.spec.ts:26:14)
    at Property.p [as predicate] (packages/fast-check/src/check/property/Property.ts:25:56)
    at Property.predicate [as run] (packages/fast-check/src/check/property/Property.generic.ts:135:27)
    at run (packages/fast-check/src/check/runner/Runner.ts:33:26)
    at runIt (packages/fast-check/src/check/runner/Runner.ts:126:7)
    at Object.check (packages/fast-check/src/check/runner/Runner.ts:178:15)
    at assert (packages/fast-check/test/e2e/NoRegressionStack.spec.ts:24:12)
    at run (packages/fast-check/test/e2e/NoRegressionStack.spec.ts:40:7)
    at _toThrowErrorMatchingSnapshot (node_modules/jest-snapshot/build/index.js:?:?)
    at Object.toThrowErrorMatchingSnapshot (node_modules/jest-snapshot/build/index.js:?:?)
    at __EXTERNAL_MATCHER_TRAP__ (node_modules/expect/build/index.js:?:?)
    at Object.throwingMatcher [as toThrowErrorMatchingSnapshot] (node_modules/expect/build/index.js:?:?)
    at Object.toThrowErrorMatchingSnapshot (packages/fast-check/test/e2e/NoRegressionStack.spec.ts:31:7)"
`;

exports[`NoRegressionStack throw 1`] = `
"Property failed after 2 tests
{ seed: 42, path: "1:0:1:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0", endOnFailure: true }
Counterexample: [0,1]
Shrunk 32 time(s)
Got error: a must be >= b
    at packages/fast-check/test/e2e/NoRegressionStack.spec.ts:12:21
    at Property.p [as predicate] (packages/fast-check/src/check/property/Property.ts:25:56)
    at Property.predicate [as run] (packages/fast-check/src/check/property/Property.generic.ts:135:27)
    at run (packages/fast-check/src/check/runner/Runner.ts:33:26)
    at runIt (packages/fast-check/src/check/runner/Runner.ts:126:7)
    at Object.check (packages/fast-check/src/check/runner/Runner.ts:178:15)
    at assert (packages/fast-check/test/e2e/NoRegressionStack.spec.ts:9:12)
    at run (packages/fast-check/test/e2e/NoRegressionStack.spec.ts:40:7)
    at _toThrowErrorMatchingSnapshot (node_modules/jest-snapshot/build/index.js:?:?)
    at Object.toThrowErrorMatchingSnapshot (node_modules/jest-snapshot/build/index.js:?:?)
    at __EXTERNAL_MATCHER_TRAP__ (node_modules/expect/build/index.js:?:?)
    at Object.throwingMatcher [as toThrowErrorMatchingSnapshot] (node_modules/expect/build/index.js:?:?)
    at Object.toThrowErrorMatchingSnapshot (packages/fast-check/test/e2e/NoRegressionStack.spec.ts:18:7)"
`;
